proposals/0000-mutable-fields.rst

```diff
@@ -0,0 +1,473 @@
+
+Operations on mutable fields
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Two modules provide APIs for operating on mutable fields::
+
+  module Data.Mutable.IO where
+    type Mutable a = Ref# RealWorld a
+    readRef :: Mutable a -> IO a
+    writeRef :: Mutable a -> a -> IO ()
+
+  module Data.Mutable.ST where
+    type Mutable s a = Ref# s a
+    readRef :: Mutable s a -> ST s a
+    writeRef :: Mutable s a -> a -> ST s ()
+
+Pattern matching
+~~~~~~~~~~~~~~~~
+
+We can pattern-match on the constructor to extract the mutable fields::
+
+  readMutPair :: MutPair a b -> IO (a, b)
+  readMutPair (MutPair aref bref) =
+    (,) <$> readRef aref <*> readRef bref
+
+When we pattern-match on a constructor with mutable fields, the
+mutable fields have type:
+
+- ``Data.Mutable.IO.Mutable a``, if the constructor has an ``IO``
+  return type
+- ``Data.Mutable.ST.Mutable s a``, if the constructor has an ``ST s``
+  return type, or a ``State# s -> (# State# s, t #)`` return type
+
+What about records?
+~~~~~~~~~~~~~~~~~~~
+
+We can use record syntax when defining the constructor::
+
+  data MutPair a b where
+    MutPair :: { mutFst :: mutable a
+               , mutSnd :: mutable b
+               } -> IO (MutPair a b}
+
+and then the record selectors are exactly what we'd expect::
+
+  mutFst :: MutPair a b -> Mutable a
+  mutSnd :: MutPair a b -> Mutable b
+
+so record selection and pattern matching work out smoothly.  Record
+construction is also fine::
+
+  do
+    mpair <- MutPair { fst = 3, snd = 4 }
+    ...
+
+but record *update* cannot be allowed for records with mutable fields,
+so GHC must reject those with an error.
+
+Primitives
+~~~~~~~~~~
+
+A mutable field is represented by a primitive type ``Ref#``::
+
+  Ref# :: * -> * -> TYPE 'Ref
+
+(``Ref`` is a new ``RuntimeRep``) and two new primitives::
+
+  readRef#  :: Ref# s a -> State# s -> (# State# s, a #)
+  writeRef# :: Ref# s a -> a -> State# s -> (# State# s, () #)
+
+A ``Ref#`` represents a mutable field of a constructor.  Although ``Ref#``
+appears to be a normal first-class primitive type, its *runtime
+representation* will be ``(# Any, Int# #)``, that is, an unboxed pair
+of the object that contains the mutable field and the offset of the
+mutable field.
+
+Unpacking constructors with mutable fields
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+While C++ or Rust allow unpacking a mutable product type into another,
+this requires a recursive notion of object construction/initialization
+and would not fit here. If we tried the following::
+
+  data MutPair2 = MP Int {-# UNPACK #-} MutPair
+
+what is the type of the MP constructor? It cannot be this::
+
+  MP :: Int -> MutPair -> IO MutPair2
+
+because the only way to make that work would be to copy the mutable
+record ``MutPair`` into the ``MP`` constructor. This is (1) not what
+we'd want for including inner objects as value types, and (2) ruins
+the guarantee that ``UNPACK`` is a performance hint rather than
+semantically important.
```

Comment:
From an imperative language background, it seems that this is what one _would_ want, at least assuming that GHC can avoid the allocation in most cases.  `UNPACK` might be the wrong pragma, though.